lib/miou_queue.ml

```diff
@@ -117,18 +116,13 @@ let iter ~f t =
   in
   go head
 
-let rec drop ~f t =
-  let head, tail = snapshot t in
-  if Atomic.compare_and_set t.head head tail then (
-    let rec go prev =
-      if prev != tail then
-        match Atomic.get prev.next with
-        | None -> ()
-        | Some next -> f next.value; go next
-    in
-    go head;
-    tail.value <- Obj.magic ())
-  else drop ~f t
+let rec drop t =
+  let ((head, tail) as snapshot) = snapshot t in
+  if Atomic.compare_and_set t.head head tail
+  then snapshot else drop t
```

Comment:
This has a space leak. The problem is that, unless the queue is already empty, the node pointed to by `tail` still has a value and after a successful `compare_and_set` that value will still be pointed to by the queue.

Reply:
Actually, that's the bug I encountered on another project. I'm not sure how to implement a atomic `drop` unfortunately. I decided to switch to a mutex + a simple queue.